db/model/Gdoc/Gdoc.ts

```diff
@@ -90,12 +98,76 @@ export class Gdoc extends BaseEntity implements OwidArticleType {
         const { text } = await gdocToArchie(data)
 
         // Convert the ArchieML to our enriched JSON structure
-        const content = archieToEnriched(text)
+        this.content = archieToEnriched(text)
+
+        this.content = await this.processContent(this.content)
 
-        this.content = content
         this.revisionId = data.revisionId ?? null
     }
 
+    // a combination of pure and impure functions where the order of operations is important
+    async processContent(
+        content: OwidArticleContent
+    ): Promise<OwidArticleContent> {
+        const links: Link[] = []
+        const allGdocs = await Gdoc.find()
+        const gdocsDictionary = keyBy(allGdocs, "id")
+        if (content.body) {
+            content.body = content.body.map((node) =>
+                recursivelyMapArticleContent(node, (node) => {
+                    // We have to track the gdoc links in use before replacing them with our slugs
+                    this.extractLink(node, links)
+                    this.replaceLink(node, gdocsDictionary)
+                    return node
+                })
+            )
+        }
+        this.links = links
+        return content
+    }
+
+    // If the node has a URL in it, create a Link object
+    // then push it into the link array from the parent closure
+    extractLink(node: OwidEnrichedArticleBlock | Span, links: Link[]): void {
+        function getText(node: OwidEnrichedArticleBlock | Span): string {
+            // Can add component-specific text accessors here
+            if (checkNodeIsSpan(node)) {
+                if (node.spanType == "span-link") {
+                    return spansToUnformattedPlainText(node.children)
+                }
+            } else if (node.type === "prominent-link") return node.title
+            return ""
+        }
+
+        if ("url" in node) {
+            const link: Link = Link.create({
+                type: getUrlType(node.url),
+                source: this,
+                target: getUrlTarget(node.url),
+                context: checkNodeIsSpan(node) ? "inline" : node.type,
+                text: getText(node),
+            })
+            links.push(link)
+        }
+    }
+
+    // Replace a gdoc link (if present) with its corresponding slug
+    // IMPURE
+    replaceLink<Node extends OwidEnrichedArticleBlock | Span>(
+        node: Node,
+        gdocsDictionary: Dictionary<Gdoc>
+    ): Node {
+        if ("url" in node && getUrlType(node.url) === "gdoc") {
+            const targetGdocId = getUrlTarget(node.url)
+            const targetGdoc = gdocsDictionary[targetGdocId]
+            if (!targetGdoc) return node
+            // TODO: handle this flow once images branch is merged
+            // throw new Error(`Document with ID ${targetGdocId} not found`)
+            node.url = "/" + targetGdoc.slug
```

Comment:
This only works on baked versions of the site, preview links won't resolve correctly.